pkg/virt-api/rest/subresource.go

```diff
@@ -787,6 +787,22 @@ func (app *SubresourceAPIApp) UnfreezeVMIRequestHandler(request *restful.Request
 
 }
 
+func (app *SubresourceAPIApp) ResetVMIRequestHandler(request *restful.Request, response *restful.Response) {
+
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if vmi.Status.Phase != v1.Running {
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf(vmNotRunning))
+		}
+		return nil
+	}
```

Comment:
What about other states like frozen or paused VMI?

Reply:
i know paused and live migration works with reset

Reply:
I think we should reset. But I thought that I read somewhere that this was at times buggy.

@berrange maybe you had metnioned this?

Reply:
Libvirt will block any operations that are considered to be unsafe/incompatible with live migration.

It permits a "reset" operation and handles this correctly. Blocking it would be pointless, as the guest OS admin can trigger a "reset" operation from the guest side at any time, and thus it *has* to be made to work regardless.

We have seen obscure bugs which only hit if a reset happens at a very specific part of migration, because it is very hard to test every conceivable code sequence here. Again though, you have to a expect a reset because the guest can trigger one at any time, and thus I see no compelling reason to block a host initiated reset.

Reply:
@davidvossel it's not a critical part of your PR, but this may be a good opportunity to avoid a race condition. Is it possible to handle errors that may happen if the VM disconnects or crashes, instead of trying to validate its state?

Reply:
> May I ask why do we need to validate the VM phase at all?


@dankenigsberg 

I don't believe we technically need to validate this to protect the system or anything. This validation seems nice though. I'd rather get an error message that the "VMI isnt' running" rather than a potentially more cryptic internal server error message from the lower layers. 

> This validation is, by nature, raceful. What happens if the validation passes but a microsecond later the VM moves to Failed or Unknown ?

In this case, the risk for a more cryptic error message would be the "race" that might occur.

The `reset` operation will still fail if we somehow make it past this point when the VMI isn't running, but the error message may get less informative as we move down the stack.

Here's the flow.

[0] client issues reset
[1] k8s apiserver
[2] kubevirt subresource endpoint
[3] virt-handler endpoint forwards to virt-launcher
[4] virt-launcher forwards to libvirt
[5] libvirt
[6] qemu


This validation is happening at step [2]. If we make it to step [3] even though the VMI is terminated, this will fail with a more generic `StatusInternalServerError` response with an error message either saying there's no VMI or there's no command socket to communicate with VMI. If we get to [4], then it will be a `StatusInternalServerError` with whatever error message libvirt gives us for the rest of the stack.

Reply:
I am guessing that [3-6] do not happen in the Context of [0]. How can the client tell that its request was unsuccessful?

It would be a nicer, raceless, API if we could report back if `reset` was successful or not, e.g by bubbling up a Condition to the vm.status. Doing so is much bigger than this PR intended to be, but would consider this?

Reply:
my 2 cents on this:
A reset is an imperative action that makes sense only if the guest is running. If the guest was just turned on/off, it becomes unreasonable and irrelevant AFAICT.

For this reason, I don't think it's valuable (and even worsens the UX) to get an answer to whether the reset operation was successful or not from the VM or another CRD, but rather it's better to get an immediate response. As @davidvossel says, even if the validation presented here will be passed unrightfully, an error from libvirt will arise which will probably not be very cryptic. In any case, even if we bubble up the error from one of our infra pods to the VM, it won't make it less cryptic.

@dankenigsberg I think that what you're suggesting here suits a declarative flow much better, since then the system would need to strive to _eventually_ fill the user's request, try again until it does so, and report back to the user about the progress. However, I don't fully understand what do we have to gain from this approach on such an imperative flow. Would you mind elaborating?

Reply:
> How can the client tell that its request was unsuccessful?

I've tried to answer it with:

> As @davidvossel says, even if the validation presented here will be passed unrightfully, an error from libvirt will arise which will probably not be very cryptic.

In other words, if the VM is shut down right after the validation takes place, the user will be able to tell that the request was unsuccessful because he would get an error from libvirt that would be bubbled up.

What am I missing?

> Do you mean that an error during [3-6] is "unreasonable and irrelevant"? 

No, that's not what I meant. I meant that if a reset operation fails because the VM was just turned on/off and the validation missed that fact because of the race, then the reset operation will fail and is probably irrelevant from this point. IOW, if you wanted to reset a VM but it was just turned off, I guess you wouldn't want to wait for it to boot and then perform the reset in an eventual consistent manner.

Reply:
@iholder101 we seem to speak past each other... If an error takes place during [0-2], the client gets a clear error. How can the client tell if an error took place during [3-6]? Where would the client get the error from libvirt?

Reply:
> How can the client tell if an error took place during [3-6]? Where would the client get the error from libvirt?

IIUC, the flow is the following:

virtctl [[1](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-fa99703f9af549f4285a734985d4f74bbfabd3a23e98ef72d1555e71702ecdb1R81)]:
```go
if err = virtClient.VirtualMachineInstance(namespace).Reset(context.Background(), vmi); err != nil {
```

virt-api [[2](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-028d148aa715883f3fd17ae07f8a03225251bdd5eb1893566dfb68085298a434R799-R801)]:
```go
	getURL := func(vmi *v1.VirtualMachineInstance, conn kubecli.VirtHandlerConn) (string, error) {
		return conn.ResetURI(vmi)
	}
```

virt-handler [[3](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-6dbc114fd509d702d01464945915b7a135d596cdc56252a8a627ea935a4075e2R573)]:
```go
ws.Route(ws.PUT("/v1/namespaces/{namespace}/virtualmachineinstances/{name}/reset").To(lifecycleHandler.ResetHandler))
```

virt-handler [[4](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-445a36c3722ec88cd4b923c17c068009eac4786851a79d727c765d3479a29b60R156)]:
```go
err = client.ResetVirtualMachine(vmi)
```

virt-launcher [[5](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-5ff91cfce1a15b5c8b1de6a76129f6b6b6eb76ee77c67f74c8941ac3c940f038R302)]:
```go
if err := l.domainManager.ResetVMI(vmi); err != nil {
```

virt-launcher [[6](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-98b8339212e612cd92ce33649e492336892a3ed649f368b181aad9aa2854138cR1777)]:
```go
if err = dom.Reset(0); err != nil {
```

And from this point onwards it will go deeper into libvirt, then to QEMU.
So the user would see the error which will be bubbled up all across the stack.

Reply:
> So the user would see the error which will be bubbled up all across the stack

Is the client request blocking until the [6] returns?

If so, https://github.com/kubevirt/kubevirt/pull/13208/files#diff-445a36c3722ec88cd4b923c17c068009eac4786851a79d727c765d3479a29b60R153 seems like a good place to respond with a race-free `VM is not running`.

Reply:
> Is the client request blocking until the [6] returns?

Yes, IIUC that's right. @davidvossel to keep me honest here.

> If so, [#13208 (files)](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-445a36c3722ec88cd4b923c17c068009eac4786851a79d727c765d3479a29b60R153) seems like a good place to respond with a race-free VM is not running.

This is possible. Let's consider this by thinking about pros/cons.

Firstly, as said above, if the VM was just turned on/off, the reset operation **should** fail, since it's not relevant anymore (I can't think why someone would want to reset a VM that had just booted for example).

Following the above, the con of what you're suggesting is that we go down the stack down to libvirt when we know that the operation will either fail or is not relevant anymore, therefore waste compute resources for no good reason. Another con is that the error message might be more cryptic when coming straight from libvirt.

When I'm trying to think about the pros here, I think the only one I can come up with is that the error will be consistent (but probably worse) in the small chance of a race. That's a stretch to even say it's a pro IMHO.

So TBH I'm not sure why removing the race in this specific case is very helpful. Can you help bring up more pros to this approach?

Reply:
> the con of what you're suggesting is that we go down the stack down to libvirt when we know that the operation will either fail or is not relevant anymore, therefore waste compute resources

Reset is a rare command. I don't think that the cost of two or three function calls would have any measurable cost in time or energy.

> Another con is that the error message might be more cryptic when coming straight from libvirt.

I have suggested to 

> https://github.com/kubevirt/kubevirt/pull/13208/files#diff-445a36c3722ec88cd4b923c17c068009eac4786851a79d727c765d3479a29b60R153 seems like a good place to respond with a race-free VM is not running.

which is the same error text.

>  I'm not sure why removing the race in this specific case is very helpful. Can you help bring up more pros to this approach?

That's the crux of the issue. Even if I cannot see an immediate risk of data loss in this case, we should consider race conditions all the time, and allow them only when

* fixing them is too complex for us
* we document them

KubeVirt is aimed to be a reliable piece of infrastructure. It should not allow race conditions just like an operating system (e.g Linux) should not. Ensuring that our APIs are race-free should be a guiding principle in everything we do.

/hold

Please unhold once the discussion about this race condition and the test code location is concluded.

Reply:
> Ensuring that our APIs are race-free should be a guiding principle in everything we do.

I respectfully disagree.
Guiding principles are great, but real life is never back and white. What we should do IMHO is assess the cons and pros on a case-to-case basis and not stick to principles as if they're strict rules. Sometimes races are perfectly reasonable. FWIW the Linux kernel has many races in cases where it makes sense.

But I believe I've expressed my opinion on this already, I'll shut up and let others comment now :)

Reply:
Just one more tiny thought

> I have suggested to
> 
> > [#13208 (files)](https://github.com/kubevirt/kubevirt/pull/13208/files#diff-445a36c3722ec88cd4b923c17c068009eac4786851a79d727c765d3479a29b60R153) seems like a good place to respond with a race-free VM is not running.

How would you respond in a race-free manner there?
Checking if the VM is running at this point is just as racy as checking it in the webhook. AFAICT this would just move the race from one place to another. The only was I see it is possible is executing the reset operation, getting an error from libvirt, and formatting it nicely. The question is whether the error message would not be cryptic enough to parse and format properly.